hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenUpdateHandler.java

```diff
@@ -374,93 +376,87 @@ private void updateKeys(
+                        requireAdminOrRole(context, originalToken, tokenKey, key);
+                    }
                 }
-
-                context.requireKey(requiredKey);
-            } else if (!isExpiryOnlyUpdateOp(op)) {
-                // if we update any other field, we need the current admin key to sign
-                // note: for expiry only op admin key is not required
-                context.requireKey(originalToken.adminKey());
             }
-        } else {
-            lowPriorityKeys = getAllRequiredLowPrioritySigners(originalToken, op);
-            final Key lowPriorityKeyList = generateKeyListKey(lowPriorityKeys);
-            context.requireKey(lowPriorityKeyList);
         }
     }
 
-    private List<Key> getAllRequiredLowPrioritySigners(
-            @NonNull final Token originalToken, @NonNull final TokenUpdateTransactionBody op) {
-        List<Key> lowPriorityKeys = new ArrayList<>();
+    private void requireAdminOrRole(
+            @NonNull final PreHandleContext context, @NonNull final Token token, @NonNull final TokenKey roleKey)
+            throws PreCheckException {
+        requireAdminOrRole(context, token, roleKey, null);
+    }
 
-        for (final var nonAdminKey : NON_ADMIN_TOKEN_KEYS) {
-            if (nonAdminKey.isPresentInUpdate(op)) {
-                final Key presentKey = nonAdminKey.getFromToken(originalToken);
-                if (presentKey != null) {
-                    lowPriorityKeys.add(presentKey);
-                }
-            }
-        }
-        // metadata field can be updated also with only metadata key signature
-        if (op.hasMetadata()) {
-            final Key metadataKey = originalToken.metadataKey();
-            if (metadataKey != null) {
-                lowPriorityKeys.add(metadataKey);
+    private void requireAdminOrRole(
+            @NonNull final PreHandleContext context,
+            @NonNull final Token token,
+            @NonNull final TokenKey roleKey,
+            @Nullable final Key replacementKey)
+            throws PreCheckException {
+        final var maybeRoleKey = roleKey.getFromToken(token);
+        mustExist(maybeRoleKey, roleKey.tokenHasNoKeyStatus());
+        if (token.hasAdminKey()) {
+            context.requireKey(oneOf(
+                    replacementKey == null ? maybeRoleKey : allOf(maybeRoleKey, replacementKey),
+                    token.adminKeyOrThrow()));
+        } else {
+            context.requireKey(maybeRoleKey);
+            if (replacementKey != null) {
+                context.requireKey(replacementKey);
             }
         }
-        return lowPriorityKeys;
     }
 
-    /**
-     * Check if TokenUpdateOp tries to remove some of the token keys
-     */
+    private void requireAdmin(@NonNull final PreHandleContext context, @NonNull final Token originalToken)
+            throws PreCheckException {
+        validateTruePreCheck(originalToken.hasAdminKey(), TOKEN_IS_IMMUTABLE);
```

Comment:
Previously, we were throwing `TOKEN_IS_IMMUTABLE` when the token cannot be updated under any circumstances. With this logic, we will throw it if the admin key is required for the success of the update transaction and it is not present. But this does not mean that the token is immutable. If the token has any low priority key, it can still be updated if for example the freeze key want to update itself

Reply:
A `TokenUpdate` that tries to change a lower-privilege key without the admin key will still succeed...it would be nice to have more targeted response codes for trying to change non-key properties without an admin key, true. 🤔